File: Sources/RedlineCore/OllamaUsage.swift
// Reads the JSONL written by the ollama shim (scripts/ollama-shim.sh). Ollama keeps no usage
// history of its own, so anything that bypasses the shim is invisible here by design.
import Foundation

/// Ollama models run in two places, and which one is a fact worth stating plainly: cloud
/// models carry a "cloud" tag in their name, everything else runs on this machine.
public enum OllamaLocality {
    public static func isCloud(_ model: String) -> Bool {
        let m = model.lowercased()
        return m.hasSuffix(":cloud") || m.hasSuffix("-cloud") || m.contains(":cloud-")
    }

    /// A cloud model gets the cloud glyph in front of its name, so the distinction is
    /// visible in every list without widening any column
    public static func marked(_ model: String) -> String {
        isCloud(model) ? "☁ \(model)" : model
    }
}

public final class OllamaStore {
    public static let provider = "Ollama"
    private let log: URL
    private let iso: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()
    private let isoPlain = ISO8601DateFormatter()

    public init(log: URL? = nil) {
        self.log = log ?? AppPaths.data("ollama.jsonl")
    }

    public var isConfigured: Bool { FileManager.default.fileExists(atPath: log.path) }

    public func scan(lookbackDays: Int, now: Date = Date()) -> [Entry] {
        guard let text = try? String(contentsOf: log, encoding: .utf8) else { return [] }
        let cutoff = now.addingTimeInterval(-Double(lookbackDays + 1) * 86400)
        var out: [Entry] = []
        text.enumerateLines { line, _ in
            guard let e = self.entry(from: line, offset: nil), e.ts > cutoff else { return }
            out.append(e)
        }
        return out
    }

    /// Reads what the shim has appended since the last pass and stores it.
    @discardableResult
    public func ingest(into warehouse: Warehouse, now: Date = Date()) -> Int {
        let path = log.path
        guard let vals = try? log.resourceValues(
            forKeys: [.contentModificationDateKey, .fileSizeKey]),
            let mtime = vals.contentModificationDate,
            let size = vals.fileSize else { return 0 }
        let mark = warehouse.ingestMark(path: path)
        let start = TranscriptTail.startOffset(mark: mark, size: size)
        if let mark, start == mark.byteOffset, size == mark.size { return 0 }

        var batch: [Entry] = []
        let next = TranscriptTail.read(url: log, from: start) { line, offset in
            guard let e = self.entry(from: line, offset: offset) else { return }
            batch.append(e)
        }
        let added = warehouse.ingest(batch)
        warehouse.setIngestMark(IngestMark(path: path, provider: OllamaStore.provider,
                                           size: size, byteOffset: next, mtime: mtime),
                                at: now)
        return added
    }

    func entry(from line: String, offset: Int?) -> Entry? {
        guard let data = line.data(using: .utf8),
              let o = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let tsStr = o["ts"] as? String,
              let ts = iso.date(from: tsStr) ?? isoPlain.date(from: tsStr) else { return nil }
        let input = int(o["prompt_eval_count"])
        let output = int(o["eval_count"])
        guard input + output > 0 else { return nil }
        // Local inference has no dollar cost, so these stay unpriced on purpose and
        // surface as token and call volume instead of spend.
        return Entry(provider: OllamaStore.provider, key: nil, ts: ts,
                     model: (o["model"] as? String) ?? "ollama",
                     input: input, output: output, cacheRead: 0, cache5m: 0, cache1h: 0,
                     origin: offset.map { "\(log.path)#\($0)" })
    }

    private func int(_ v: Any?) -> Int {
        if let i = v as? Int { return i }
        if let d = v as? Double { return Int(d) }
        return 0
    }
}

File: windows/src/Redline.Core/OllamaUsage.cs
// Reads the JSONL written by the ollama shim (`redlinectl ollama`). Ollama keeps no usage
// history of its own, so anything that bypasses the shim is invisible here by design.
namespace Redline.Core;

/// Cloud models carry a "cloud" tag in their name; everything else runs on this machine.
public static class OllamaLocality
{
    public static bool IsCloud(string model)
    {
        var m = model.ToLowerInvariant();
        return m.EndsWith(":cloud", StringComparison.Ordinal) || m.EndsWith("-cloud", StringComparison.Ordinal) ||
               m.Contains(":cloud-", StringComparison.Ordinal);
    }

    /// A cloud model gets the cloud glyph in front of its name, visible in every list at no width.
    public static string Marked(string model) => IsCloud(model) ? "☁ " + model : model;
}

public sealed class OllamaStore
{
    public const string Provider = "Ollama";
    private readonly string _log;

    public OllamaStore(string? log = null)
    {
        _log = log ?? RedlineHome.PathFor(".local/share/redline/ollama.jsonl");
    }

    public bool IsConfigured => File.Exists(_log);

    public List<Entry> Scan(int lookbackDays, DateTimeOffset? now = null)
    {
        var output = new List<Entry>();
        string text;
        try { text = File.ReadAllText(_log); } catch { return output; }
        var cutoff = (now ?? DateTimeOffset.UtcNow).AddSeconds(-(double)(lookbackDays + 1) * 86400);
        using var reader = new StringReader(text);
        string? line;
        while ((line = reader.ReadLine()) is not null)
        {
            if (EntryFrom(line, null) is { } e && e.Ts > cutoff) output.Add(e);
        }
        return output;
    }

    /// Reads what the shim has appended since the last pass and stores it.
    public int Ingest(Warehouse warehouse, DateTimeOffset? now = null)
    {
        var at = now ?? DateTimeOffset.UtcNow;
        FileInfo info;
        try { info = new FileInfo(_log); if (!info.Exists) return 0; }
        catch { return 0; }
        var mtime = new DateTimeOffset(info.LastWriteTimeUtc, TimeSpan.Zero);
        var size = info.Length;
        var mark = warehouse.IngestMark(_log);
        var start = TranscriptTail.StartOffset(mark, size);
        if (mark is not null && start == mark.ByteOffset && size == mark.Size) return 0;

        var batch = new List<Entry>();
        var next = TranscriptTail.Read(_log, start, (line, offset) =>
        {
            if (EntryFrom(line, offset) is { } e) batch.Add(e);
        });
        var added = warehouse.Ingest(batch);
        warehouse.SetIngestMark(new IngestMark(_log, Provider, size, next, mtime), at);
        return added;
    }

    internal Entry? EntryFrom(string line, long? offset)
    {
        if (Json.ParseObject(line) is not { } o ||
            TranscriptTail.ParseTimestamp(Json.Str(o["ts"])) is not { } ts) return null;
        var input = Int(o["prompt_eval_count"]);
        var output = Int(o["eval_count"]);
        if (input + output <= 0) return null;
        // Local inference has no dollar cost, so these stay unpriced and surface as volume, not spend
        return new Entry(Provider, null, ts, Json.Str(o["model"]) ?? "ollama",
                         input, output, 0, 0, 0, offset is { } off ? $"{_log}#{off}" : null);
    }

    private static int Int(System.Text.Json.Nodes.JsonNode? v) => Json.Num(v) is { } d ? (int)d : 0;
}
